File: tutorials/python/hello_world/server/helloworldpeer.py
from spa.serverside import CClientPeer
from spa import CScopeUQueue as Sb, CServerError as Se
from msstruct import CMyStruct
import time


class CHelloWorldPeer(CClientPeer):

    def sayHello(self):
        fName = self.UQueue.LoadString()
        if not fName or not len(fName):
            raise Se(12345, 'First name cannot be empty!')
        lName = self.UQueue.LoadString()
        res = u'Hello ' + fName + ' ' + lName
        print(res)
        return Sb().SaveString(res)

    def sleep(self):
        ms = self.UQueue.LoadInt()
        if ms < 0:
            raise Se(54321, 'Sleep time cannot be less than zero')
        time.sleep(ms/1000.0)

    def echo(self):
        # return self.UQueue #this also works
        ms = CMyStruct()
        ms.LoadFrom(self.UQueue)
        return ms.SaveTo(Sb())

File: tutorials/java/src/hello_world/server/HelloWorldPeer.java
package hello_world.server;

import SPA.ServerSide.*;
import hello_world.hwConst;
import uqueue_demo.CMyStruct;
import SPA.CServerError;

public class HelloWorldPeer extends CClientPeer {

    @RequestAttr(RequestID = hwConst.idSayHello)
    private String SayHello(String firstName, String lastName) throws CServerError {
        if (firstName == null || firstName.length() == 0) {
            throw new SPA.CServerError(62345, "First name cannot be empty");
        }
        String res = "Hello " + firstName + " " + lastName;
        System.out.println(res);
        return res;
    }

    @RequestAttr(RequestID = hwConst.idSleep, SlowRequest = true) //true -- slow request
    private void Sleep(int ms) throws CServerError {
        if (ms < 0) {
            throw new SPA.CServerError(12345, "ms cannot be negative");
        }
        try {
            Thread.sleep(ms);
        } catch (InterruptedException err) {
        }
    }

    @RequestAttr(RequestID = hwConst.idEcho)
    private CMyStruct Echo(CMyStruct ms) {
        return ms;
    }
}
